src/java.base/share/classes/jdk/internal/classfile/AttributedElement.java

```diff
@@ -0,0 +1,108 @@
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.classfile;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import jdk.internal.classfile.attribute.RecordComponentInfo;
+import jdk.internal.classfile.impl.AbstractUnboundModel;
+
+/**
+ * A {@link ClassfileElement} describing an entity that has attributes, such
+ * as a class, field, method, code attribute, or record component.
+ */
+public sealed interface AttributedElement extends ClassfileElement
+        permits ClassModel, CodeModel, FieldModel, MethodModel,
+                RecordComponentInfo, AbstractUnboundModel {
+
+    /**
+     * {@return the kind of the attributed element}
+     */
+    Kind attributedElementKind();
+
+    /**
+     * {@return the attributes of this element}
+     */
+    List<Attribute<?>> attributes();
+
+    /**
+     * Finds an attribute by name.
+     * @param attr the attribute mapper
+     * @param <T> the type of the attribute
+     * @return the attribute, or an empty {@linkplain Optional} if the attribute
+     * is not present
+     */
+    default <T extends Attribute<T>> Optional<T> findAttribute(AttributeMapper<T> attr) {
+        for (Attribute<?> la : attributes()) {
+            if (la.attributeMapper() == attr) {
+                @SuppressWarnings("unchecked")
+                var res = Optional.of((T) la);
+                return res;
+            }
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Finds one or more attributes by name.
+     * @param attr the attribute mapper
+     * @param <T> the type of the attribute
+     * @return the attributes, or an empty {@linkplain List} if the attribute
+     * is not present
+     */
+    default <T extends Attribute<T>> List<T> findAttributes(AttributeMapper<T> attr) {
+        var list = new ArrayList<T>();
+        for (var a : attributes()) {
+            if (a.attributeMapper() == attr) {
+                @SuppressWarnings("unchecked")
+                T t = (T)a;
+                list.add(t);
+            }
+        }
+        return list;
+    }
+
+    /**
+     * Enum constants describing locations in the classfile where attributes
+     * are permitted.
+     */
+    enum Kind {
```

Comment:
There are at least 72 usages of AttributedElement.Kind across the Classfile API.
Do you suggest to rename it to something else (for example Location)?

Reply:
Uhm - I can't see these usages... something seems to be off with my IDE configuration. I did a grep and I now saw the uses. That said, having the Kind/Location inside AttributedElement still looks weird to me. The "places where an attribute can appear" is a property of an `Attribute`, not of an attributed element (which is used to model elements which can have attributes).

Reply:
Still, there seems to be a modelling issue here. The property of "where could this attribute go" is a property of the attribute. Of course, for usability reason, an AttributedElement might expose a predicate saying "I only accept attributes of these kinds". But it seems to me as if the API has this relationship backwards, due to _where_ the Kind interface is being defined. I think if `Kind` was defined in `AttributeMapper` it would be a tad easier to understand. And, perhaps, the `attributedElementKind` name should be changed to something like `applicableAttributeKinds` or something like that.

Reply:
The relation is 1 `Attribute` is applicable in N `AttributedElements` and not vice versa.
For example `ClassModel::attributedElementKind` returns `CLASS` and for example `Attributes.RUNTIME_INVISIBLE_TYPE_ANNOTATIONS::applicableAttributeKinds` returns `Set.of(CLASS, METHOD, FIELD, CODE_ATTRIBUTE, RECORD_COMPONENT)`, so we know it is applicable.

However I'll try to re-visit this part of API if needed at all.

Reply:
I understand what the API, in its current form, wants to do. IMHO, the same functionality would be better expressed if the API had a way to say:

* each attribute has a _target_, where target can be a set of CLASS, METHOD, ...
* an attribute mapper supports a set of attribute targets
* an attributed element also supports a set of attribute targets

E.g. the target of the attribute is the concept to model - then AttributeMapper and AttributedElement just happen to expose a set of supported targets, which clients can query if needed.

Reply:
I've prepared pull request to jdk-sandbox with complete removal of the AttributedElement.Kind and sent following note to classfile-api-dev at openjdk.org :


I would like to propose removal of AttributedElement.Kind across all Classfile API.
The AttributedElement.Kind models Attributes “where applicable” and it is a duplication of each Attribute extending ClassElement, MethodElement, CodeElement, etc…

Classfile API is not actively using the AttributedElement.Kind except for parsing, where inappropriate AttributedElement.Kind is resolved as UnknownAttribute.

Following proposal removes all usages of AttributedElement.Kind from Classfile API:
https://github.com/openjdk/jdk-sandbox/pull/48/files

Please let me know is there are any objections.